src/clients/avail/client.rs

```diff
@@ -0,0 +1,249 @@
+        bytes32 blobRoot;
+        // bridge root to check proof against, or reconstruct the data root
+        bytes32 bridgeRoot;
+        // leaf being proven
+        bytes32 leaf;
+        // index of the leaf in the blob/bridge root tree
+        uint256 leafIndex;
+    }
+}
+
+impl AvailClient {
+    pub async fn new() -> anyhow::Result<Self> {
+        let config = AvailConfig::from_env()?;
+
+        let client = AvailSubxtClient::new(config.api_node_url.clone())
+            .await
+            .map_err(|e| types::DAError {
+                error: e.into(),
+                is_retriable: false,
+            })?;
+
+        let mnemonic = Mnemonic::parse(&config.seed).map_err(|e| types::DAError {
+            error: e.into(),
+            is_retriable: false,
+        })?;
+
+        let keypair = Keypair::from_phrase(&mnemonic, None).map_err(|e| types::DAError {
+            error: e.into(),
+            is_retriable: false,
+        })?;
+
+        let api_client = reqwest::Client::new();
+
+        Ok(Self {
+            config,
+            client: client.into(),
+            api_client: api_client.into(),
+            keypair,
+        })
+    }
+
+    fn to_non_retriable_da_error(&self, error: anyhow::Error) -> types::DAError {
+        DAError {
+            error,
+            is_retriable: false,
+        }
+    }
+}
+
+#[async_trait]
+impl DataAvailabilityClient for AvailClient {
+    async fn dispatch_blob(
+        &self,
+        _batch_number: u32,
+        data: Vec<u8>,
+    ) -> Result<DispatchResponse, DAError> {
+        let call = api::tx()
+            .data_availability()
+            .submit_data(BoundedVec(data.clone()));
+        let tx_progress = tx::send(
+            &self.client,
+            &call,
+            &self.keypair,
+            AppId(self.config.app_id),
+        )
+        .await
+        .map_err(|e| self.to_non_retriable_da_error(e.into()))?;
+        let block_hash = tx::then_in_block(tx_progress)
+            .await
+            .map_err(|e| self.to_non_retriable_da_error(e.into()))?
+            .block_hash();
+
+        // Retrieve the data from the block hash
+        let block = self
+            .client
+            .blocks()
+            .at(block_hash)
+            .await
+            .map_err(|e| self.to_non_retriable_da_error(e.into()))?;
+        let extrinsics = block
+            .extrinsics()
+            .await
+            .map_err(|e| self.to_non_retriable_da_error(e.into()))?;
+        let mut found = false;
+        let mut tx_idx = 0;
+        for ext in extrinsics.iter() {
+            let ext = ext.map_err(|e| self.to_non_retriable_da_error(e.into()))?;
+            let call = ext.as_extrinsic::<SubmitData>();
+            if let Ok(Some(call)) = call {
+                if data.clone() == call.data.0 {
+                    found = true;
+                    break;
+                }
+            }
+            tx_idx += 1;
+        }
```

Comment:
This code looks kinda hacky, is it possible to use e.g. [`TxInBlock::extrinsic_hash`]( https://docs.rs/subxt/latest/subxt/tx/struct.TxInBlock.html#method.extrinsic_hash) instead of `tx_idx`?

Reply:
Unfortunately not, two issues are there:
* There is no good way to fetch an extrinsic for an extrinsic hash, and hashes are not unique
* The bridge API relies on a block hash and tx index for giving us DA inclusion proofs